src/content/en/updates/posts/2016/03/web-enabled-internet-of-things.markdown

```diff
@@ -0,0 +1,523 @@
+Recall in the notes above that `eddystone-beacon` will not work
+with bleno unless you use the noble version distributed with it?
+Doing that is luckily quite simple:
+
+{% highlight javascript %}
+var bleno = require('eddystone-beacon/node_modules/bleno');
+var util = require('util');
+{% endhighlight %}
+
+Now all we need it to advertise our device (UUID) and its
+characteristics (other UUIDs)
+
+{% highlight javascript %}
+bleno.on('advertisingStart', function(error) {
+    ...
+    bleno.setServices([
+      new bleno.PrimaryService({
+        uuid: 'fc00',
+        characteristics: [
+          temperatureCharacteristic, colorCharacteristic
+        ]
+      })
+    ]);
+});
+{% endhighlight %}
+
+# Creating the client web app
+
+I don't want to go into too much details on how the non-bluetooth
+parts of the client app works, but I can tell you that I have used
+Polymer for easily creating a responsive user interface which serves
+the purpose of this demo. The resulting app looks like the below:
+
+<img alt="Client app on phone" src="/web/updates/images/2016/03/web-enabled-internet-of-things/phone-ux.png" style="width: 50%; float: left"/>
+<img alt="Error message" src="/web/updates/images/2016/03/web-enabled-internet-of-things/error.png" style="width: 50%; float: left"/>
+<br/>
+
+The right side shows an earlier version, show casing a simple error
+log that I added, easing the development.
+
+Web Bluetooth makes it dead easy to talk to a Bluetooth Low Energy
+device, so let's look at a simplified version of my connection code.
+If you don't know how promises work, please check out this resource
+before reading further:
+
+https://developers.google.com/web/fundamentals/primers/promises
+
+Connecting to a Bluetooth device involved a chain of promises.
+First we filter for the device (UUID: `FC00`, name: `Edison`), which
+will popup a dialog for the user to select the device given the
+filter. Then we will connect to the GATT service and get the primary
+service and then the characteristics, which we then act on (read
+values and set up notification callbacks).
+
+> NOTE: It is best practise to **not** fetch the characteristics
+> *in parallel* using something like `Promise.all([p1, p2])`, as that
+> will make sure the successive read/writes following in the promise
+> chain will happen properly.
+
+Below we show a simplified version of my code, which will only work
+with the latest Web Bluetooth API, thus requiring Chrome Dev on Android.
+
+{% highlight javascript %}
+navigator.bluetooth.requestDevice({
+  filters: [{ name: 'Edison' }],
+  optionalServices: [0xFC00]
+})
+
+.then((device) => {
+  return device.gatt.connect();
+})
+
+.then((server) => {
+  return server.getPrimaryService(0xFC00));
+});
+
+.then((service) => {
+  let p1 = () => service.getCharacteristic(0xFC0B)
+  .then((characteristic) => {
+    this.colorLedCharacteristic = characteristic;
+    this.readLedColor();
```

Comment:
You need to return the promise otherwise the `startNotifications()` could could be executed at the same time as `readValue()` and one of the would fail.